### PR TITLE
consensus: only call privValidator.GetPubKey once per block

### LINKED ISSUE #4865: Excessive calls to GetPubkey() on external signer
**Tendermint version** 
All, including master

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
gaiad

**Environment**:
- **OS** (e.g. from /etc/os-release): Ubuntu 

**What happened**:
When using an external KMS, GetPubkey() is called a minimum of eight times per block. As the pubkey can reasonably be expected to not change during a given block, this should be cached on the Tendermint side of the connection, rather than pushing packets back and forth over the connection (plus marshalling / unmarshalling overhead - which I appreciate is less now than it previously was). It will also reduce the number of calls to the HSM (where such caching isn't implemented at the KMS) and allow for more concurrent chains to be run on a single HSM before hitting capacity.

**What you expected to happen**:
GetPubkey gets called once, and once only, per block.

**Have you tried the latest version**: yes/no
yes

[edit: once per block, rather than once; thanks @melekes ]

### PR DESCRIPTION
**Pull request by melekes**
Closes #4865

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -262,11 +268,17 @@ func (cs *State) GetValidators() (int64, []*types.Validator) {
...
 func (cs *State) SetPrivValidator(priv types.PrivValidator) {
 	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
 	cs.privValidator = priv
-	cs.mtx.Unlock()
+
+	if err := cs.updatePrivValidatorPubKey(); err != nil {
+		cs.Logger.Error("Can't get private validator pubkey", "err", err)
+	}
```
Comment (erikgrinaker) [inline consensus/state.go:281]: Should this be a hard error condition? If we can't fetch the validator's public key on startup, is there even any point in continuing with consensus?

Comment (melekes) [reply consensus/state.go:281]: the very first time we try to get pubkey is in node.go https://github.com/tendermint/tendermint/blob/48426c5992c3f17c80f8c23fff84640b9f03c997/node/node.go#L664

it's fine to continue here => validator will just not sign / vote on this block (or more depending on downtime)

**Comment by melekes** [comment]
> There is another key cache in the metrics function could this be replaced with this change?

I can't use `privValidatorPubKey` because `privValidator` might be nil, so we need this "cache" in any case.

> Also is there any chance we could test this and see the improvement?

you mean write a test or test by hands?

**Comment by marbar3778** [comment]
> you mean write a test or test by hands?

test by hand. I dont think we have e2e tests that could do this, do we?

**Comment by marbar3778** [comment]
is this PR good to go into master?

**Pull request merged**